combinetf2/fitter.py

```diff
@@ -0,0 +1,1410 @@
+                    shape=(impacts.shape[0],), dtype=tf.float64
+                ),
+            )
+            impacts_grouped_syst = tf.transpose(impacts_grouped_syst)
+            impacts_grouped = tf.concat([impacts_grouped_syst, impacts_grouped], axis=1)
+
+        return impacts, impacts_grouped
+
+    def _compute_global_impact_group(self, d_squared, idxs):
+        gathered = tf.gather(d_squared, idxs, axis=-1)
+        d_squared_summed = tf.reduce_sum(gathered, axis=-1)
+        return tf.sqrt(d_squared_summed)
+
+    @tf.function
+    def global_impacts_parms(self):
+        # compute impacts for pois and nois
+        dxdtheta0, dxdnobs, dxdbeta0 = self._compute_derivatives_x()
+
+        dxdtheta0_poi = dxdtheta0[: self.npoi]
+        dxdtheta0_noi = tf.gather(dxdtheta0[self.npoi :], self.indata.noigroupidxs)
+        dxdtheta0 = tf.concat([dxdtheta0_poi, dxdtheta0_noi], axis=0)
+        dxdtheta0_squared = tf.square(dxdtheta0)
+
+        # global impact data stat
+        dxdnobs_poi = dxdnobs[: self.npoi]
+        dxdnobs_noi = tf.gather(dxdnobs[self.npoi :], self.indata.noigroupidxs)
+        dxdnobs = tf.concat([dxdnobs_poi, dxdnobs_noi], axis=0)
+
+        if self.externalCovariance:
+            data_cov = tf.linalg.inv(self.data_cov_inv)
+            data_stat = tf.einsum("ij,jk,ik->i", dxdnobs, data_cov, dxdnobs)
+        else:
+            data_stat = tf.reduce_sum(tf.square(dxdnobs) * self.nobs, axis=-1)
+
+        data_stat = tf.sqrt(data_stat)
+        impacts_data_stat = tf.reshape(data_stat, (-1, 1))
+
+        if self.binByBinStat:
+            # global impact bin-by-bin stat
+            dxdbeta0_poi = dxdbeta0[: self.npoi]
+            dxdbeta0_noi = tf.gather(dxdbeta0[self.npoi :], self.indata.noigroupidxs)
+            dxdbeta0 = tf.concat([dxdbeta0_poi, dxdbeta0_noi], axis=0)
+
+            impacts_bbb = tf.sqrt(
+                tf.reduce_sum(
+                    tf.square(dxdbeta0) * tf.math.reciprocal(self.indata.kstat), axis=-1
+                )
+            )
+            impacts_bbb = tf.reshape(impacts_bbb, (-1, 1))
+            impacts_grouped = tf.concat([impacts_data_stat, impacts_bbb], axis=1)
+        else:
+            impacts_grouped = impacts_data_stat
+
+        if len(self.indata.systgroupidxs):
+            impacts_grouped_syst = tf.map_fn(
+                lambda idxs: self._compute_global_impact_group(dxdtheta0_squared, idxs),
+                tf.ragged.constant(self.indata.systgroupidxs, dtype=tf.int32),
+                fn_output_signature=tf.TensorSpec(
+                    shape=(dxdtheta0_squared.shape[0],), dtype=tf.float64
+                ),
+            )
+            impacts_grouped_syst = tf.transpose(impacts_grouped_syst)
+            impacts_grouped = tf.concat([impacts_grouped_syst, impacts_grouped], axis=1)
+
+        # global impacts of unconstrained parameters are always 0, only store impacts of constrained ones
+        impacts = dxdtheta0[:, self.indata.nsystnoconstraint :]
```

Comment:
Even if they're always zero doesn't it make sense to store it for the sake of keeping all lists of parameters consistent?